File: 2025/day11/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Graph map[string][]string

func parseInput(filename string) (Graph, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	graph := make(Graph)
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		parts := strings.Split(line, ":")
		if len(parts) != 2 {
			continue
		}

		source := strings.TrimSpace(parts[0])
		destinations := strings.Fields(strings.TrimSpace(parts[1]))

		graph[source] = destinations
	}

	return graph, scanner.Err()
}

func countPaths(graph Graph, current string, target string, visited map[string]bool) int {
	// Base case: reached target
	if current == target {
		return 1
	}

	// Mark current node as visited
	visited[current] = true
	defer func() { delete(visited, current) }() // Unmark when backtracking

	// Get neighbors of current node
	neighbors, exists := graph[current]
	if !exists {
		return 0
	}

	totalPaths := 0
	for _, neighbor := range neighbors {
		// Skip if already visited (avoid cycles)
		if visited[neighbor] {
			continue
		}

		// Recursively count paths from neighbor to target
		totalPaths += countPaths(graph, neighbor, target, visited)
	}

	return totalPaths
}

func main() {
	graph, err := parseInput("input.txt")
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	visited := make(map[string]bool)
	pathCount := countPaths(graph, "you", "out", visited)

	fmt.Printf("Number of paths from 'you' to 'out': %d\n", pathCount)
}

File: 2025/day11/main.py
#!/usr/bin/env python3

def parse_input(filename):
    """Parse the input file and build a graph as an adjacency list."""
    graph = {}

    with open(filename, 'r') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue

            # Split by colon
            parts = line.split(':')
            if len(parts) != 2:
                continue

            source = parts[0].strip()
            destinations = parts[1].strip().split()

            graph[source] = destinations

    return graph

def count_paths(graph, current, target, visited):
    """
    Count all paths from current node to target node using DFS.

    Args:
        graph: Adjacency list representation of the graph
        current: Current node name
        target: Target node name
        visited: Set of visited nodes to avoid cycles

    Returns:
        Number of paths from current to target
    """
    # Base case: reached target
    if current == target:
        return 1

    # Mark current node as visited
    visited.add(current)

    # Get neighbors of current node
    neighbors = graph.get(current, [])

    total_paths = 0
    for neighbor in neighbors:
        # Skip if already visited (avoid cycles)
        if neighbor in visited:
            continue

        # Recursively count paths from neighbor to target
        total_paths += count_paths(graph, neighbor, target, visited)

    # Unmark when backtracking
    visited.remove(current)

    return total_paths

def main():
    # Parse the input
    graph = parse_input("input.txt")

    print(f"Parsed {len(graph)} devices")

    # Count paths from 'you' to 'out'
    visited = set()
    path_count = count_paths(graph, "you", "out", visited)

    print(f"Number of paths from 'you' to 'out': {path_count}")

if __name__ == "__main__":
    main()
